naga/src/proc/constant_evaluator.rs

```diff
@@ -254,6 +254,256 @@ gen_component_wise_extractor! {
     ],
 }
 
+/// Vector for each [`Literal`] type
+///
+/// This type ensures that all elements have same type
+#[derive(Debug)]
+enum LiteralVector {
+    F64(ArrayVec<f64, { crate::VectorSize::MAX }>),
+    F32(ArrayVec<f32, { crate::VectorSize::MAX }>),
+    U32(ArrayVec<u32, { crate::VectorSize::MAX }>),
+    I32(ArrayVec<i32, { crate::VectorSize::MAX }>),
+    U64(ArrayVec<u64, { crate::VectorSize::MAX }>),
+    I64(ArrayVec<i64, { crate::VectorSize::MAX }>),
+    Bool(ArrayVec<bool, { crate::VectorSize::MAX }>),
+    AbstractInt(ArrayVec<i64, { crate::VectorSize::MAX }>),
+    AbstractFloat(ArrayVec<f64, { crate::VectorSize::MAX }>),
+}
+
+impl LiteralVector {
+    #[allow(clippy::pattern_type_mismatch, clippy::missing_const_for_fn)]
+    fn len(&self) -> usize {
+        match self {
+            LiteralVector::F64(v) => v.len(),
+            LiteralVector::F32(v) => v.len(),
+            LiteralVector::U32(v) => v.len(),
+            LiteralVector::I32(v) => v.len(),
+            LiteralVector::U64(v) => v.len(),
+            LiteralVector::I64(v) => v.len(),
+            LiteralVector::Bool(v) => v.len(),
+            LiteralVector::AbstractInt(v) => v.len(),
+            LiteralVector::AbstractFloat(v) => v.len(),
+        }
+    }
+    /// Creates [`LiteralVector`] of size 1 from single [`Literal`]
+    fn from_literal(literal: Literal) -> Self {
+        match literal {
+            Literal::F64(e) => Self::F64(ArrayVec::from_iter(iter::once(e))),
+            Literal::F32(e) => Self::F32(ArrayVec::from_iter(iter::once(e))),
+            Literal::U32(e) => Self::U32(ArrayVec::from_iter(iter::once(e))),
+            Literal::I32(e) => Self::I32(ArrayVec::from_iter(iter::once(e))),
+            Literal::U64(e) => Self::U64(ArrayVec::from_iter(iter::once(e))),
+            Literal::I64(e) => Self::I64(ArrayVec::from_iter(iter::once(e))),
+            Literal::Bool(e) => Self::Bool(ArrayVec::from_iter(iter::once(e))),
+            Literal::AbstractInt(e) => Self::AbstractInt(ArrayVec::from_iter(iter::once(e))),
+            Literal::AbstractFloat(e) => Self::AbstractFloat(ArrayVec::from_iter(iter::once(e))),
+        }
```

Comment:
nitpick: This code might benefit from a local `fn` here that can capture the `ArrayVec::from_iter(iter::once(…))` detail, but I don't feel strongly about changing it.